It works in the test, but when I try to use in my application, I get: 
```
Exception in thread "main" java.lang.ExceptionInInitializerError
	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl1.<clinit>(Unknown Source)
	... 1 more
Caused by: java.lang.NoClassDefFoundError: io/quarkus/undertow/deployment/UndertowHandlersConfServletExtension
	at io.quarkus.deployment.steps.UndertowBuildStep$build13.deploy(Unknown Source)
	... 2 more
Caused by: java.lang.ClassNotFoundException: io.quarkus.undertow.deployment.UndertowHandlersConfServletExtension
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 3 more
```

Any thoughts? @stuartwdouglas ?

--- a/docs/src/main/asciidoc/index.adoc
+++ b/docs/src/main/asciidoc/index.adoc
@@ -35,6 +35,7 @@ include::quarkus-intro.adoc[tag=intro]
 * link:health-guide.html[Using MicroProfile Health]
 * link:fault-tolerance-guide.html[Using Fault Tolerance]
 * link:context-propagation-guide.html[Context Propagation]
+* link:undertow-reference.html[Undertow Reference]
 * link:extension-authors-guide.html[Write Your Own Extension] _(advanced)_
 * link:writing-native-applications-tips.html[Tips for writing native applications] _(advanced)_
 * link:performance-measure.html[Measuring Performance] _(advanced)_

--- a/docs/src/main/asciidoc/undertow-reference.adoc
+++ b/docs/src/main/asciidoc/undertow-reference.adoc
@@ -1,0 +1,32 @@
+////
+This guide is maintained in the main Quarkus repository
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
+////
+
+include::./attributes.adoc[]
+= {project-name} - Undertow
+:numbered:
+:sectnums:
+:sectnumlevels: 4
+:toc:
+
+:numbered:
+:sectnums:
+:sectnumlevels: 4
+
+
+This document explains various Undertow features that you can use in Quarkus.
+
+
+=== Serving Static Resources
+
+To serve static resources you must place them in the `META-INF/resources` directory of your application. This is because
+Quarkus applications are jar files rather than war files, and this location is the standard location for resources in
+jar files as defined by the Servlet spec.
+
+=== undertow-handlers.conf
+
+You can make use of the Undertow predicate language using an `undertow-handlers.conf` file. This file should be placed
+in the `META-INF` directory of your application jar. This file contains handlers defined using the
+link:http://undertow.io/undertow-docs/undertow-docs-2.0.0/index.html#predicates-attributes-and-handlers[Undertow predicate language].

--- a/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java
+++ b/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java
@@ -116,6 +116,7 @@ import io.quarkus.undertow.runtime.ServletSecurityInfoProxy;
 import io.quarkus.undertow.runtime.ServletSecurityInfoSubstitution;
 import io.quarkus.undertow.runtime.UndertowDeploymentTemplate;
 import io.quarkus.undertow.runtime.filters.CORSTemplate;
+import io.quarkus.undertow.runtime.UndertowHandlersConfServletExtension;
 import io.undertow.Undertow;
 import io.undertow.servlet.api.DeploymentInfo;
 import io.undertow.servlet.api.FilterInfo;
@@ -198,6 +199,17 @@ public class UndertowBuildStep {
     @BuildStep
     public void kubernetes(HttpConfig config, BuildProducer<KubernetesPortBuildItem> portProducer) {
         portProducer.produce(new KubernetesPortBuildItem(config.port, "http"));
+    }
+
+    /**
+     * Register the undertow-handlers.conf file
+     */
+    @BuildStep
+    @Record(STATIC_INIT)
+    public void registerUndertowHandlersConf(BuildProducer<ServletExtensionBuildItem> producer) {
+        if (UndertowHandlersConfServletExtension.existsConfFile()) {
+            producer.produce(new ServletExtensionBuildItem(new UndertowHandlersConfServletExtension()));
+        }
     }
 
     @Record(STATIC_INIT)

--- a/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java
+++ b/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java
@@ -97,6 +97,7 @@ import io.quarkus.deployment.annotations.Record;
 import io.quarkus.deployment.builditem.ApplicationArchivesBuildItem;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.ExecutorBuildItem;
+import io.quarkus.deployment.builditem.HotDeploymentWatchedFileBuildItem;
 import io.quarkus.deployment.builditem.LaunchModeBuildItem;
 import io.quarkus.deployment.builditem.ObjectSubstitutionBuildItem;
 import io.quarkus.deployment.builditem.ServiceStartBuildItem;
@@ -115,8 +116,8 @@ import io.quarkus.undertow.runtime.ServletProducer;
 import io.quarkus.undertow.runtime.ServletSecurityInfoProxy;
 import io.quarkus.undertow.runtime.ServletSecurityInfoSubstitution;
 import io.quarkus.undertow.runtime.UndertowDeploymentTemplate;
-import io.quarkus.undertow.runtime.filters.CORSTemplate;
 import io.quarkus.undertow.runtime.UndertowHandlersConfServletExtension;
+import io.quarkus.undertow.runtime.filters.CORSTemplate;
 import io.undertow.Undertow;
 import io.undertow.servlet.api.DeploymentInfo;
 import io.undertow.servlet.api.FilterInfo;
@@ -206,9 +207,21 @@ public class UndertowBuildStep {
      */
     @BuildStep
     @Record(STATIC_INIT)
-    public void registerUndertowHandlersConf(BuildProducer<ServletExtensionBuildItem> producer) {
-        if (UndertowHandlersConfServletExtension.existsConfFile()) {
+    public void registerUndertowHandlersConf(BuildProducer<ServletExtensionBuildItem> producer,
+            ApplicationArchivesBuildItem applicationArchivesBuildItem,
+            BuildProducer<HotDeploymentWatchedFileBuildItem> watchedFile,
+            BuildProducer<SubstrateResourceBuildItem> substrateResourceBuildItemBuildProducer) {
+        //we always watch the file, so if it gets added we restart
+        watchedFile.produce(
+                new HotDeploymentWatchedFileBuildItem(UndertowHandlersConfServletExtension.META_INF_UNDERTOW_HANDLERS_CONF));
+
+        //check for the file in the handlers dir
+        Path handlerPath = applicationArchivesBuildItem.getRootArchive()
+                .getChildPath(UndertowHandlersConfServletExtension.META_INF_UNDERTOW_HANDLERS_CONF);
+        if (handlerPath != null) {
             producer.produce(new ServletExtensionBuildItem(new UndertowHandlersConfServletExtension()));
+            substrateResourceBuildItemBuildProducer.produce(
+                    new SubstrateResourceBuildItem(UndertowHandlersConfServletExtension.META_INF_UNDERTOW_HANDLERS_CONF));
         }
     }
 

--- a/extensions/undertow/deployment/src/test/java/io/quarkus/undertow/test/UndertowHandlersConfTestCase.java
+++ b/extensions/undertow/deployment/src/test/java/io/quarkus/undertow/test/UndertowHandlersConfTestCase.java
@@ -1,0 +1,45 @@
+package io.quarkus.undertow.test;
+
+import static io.restassured.RestAssured.when;
+import static org.hamcrest.Matchers.is;
+
+import java.io.IOException;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class UndertowHandlersConfTestCase {
+
+    @RegisterExtension
+    static QuarkusUnitTest runner = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(RedirectedServlet.class)
+                    .addAsManifestResource(new StringAsset("path-prefix('/foo') -> rewrite('/bar')"),
+                            "undertow-handlers.conf"));
+
+    @Test
+    public void testUndertowHandlersRewrite() {
+        when().get("/foo").then()
+                .statusCode(200)
+                .body(is("RedirectedServlet"));
+    }
+
+    @WebServlet(urlPatterns = "/bar")
+    public static class RedirectedServlet extends HttpServlet {
+        @Override
+        protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+            resp.getWriter().print("RedirectedServlet");
+        }
+    }
+
+}

--- a/extensions/undertow/runtime/src/main/java/io/quarkus/undertow/runtime/RewriteCorrectingHandlerWrappers.java
+++ b/extensions/undertow/runtime/src/main/java/io/quarkus/undertow/runtime/RewriteCorrectingHandlerWrappers.java
@@ -1,0 +1,57 @@
+package io.quarkus.undertow.runtime;
+
+import io.undertow.server.HandlerWrapper;
+import io.undertow.server.HttpHandler;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.servlet.handlers.ServletPathMatch;
+import io.undertow.servlet.handlers.ServletRequestContext;
+import io.undertow.util.AttachmentKey;
+
+/**
+ * Handler that works around issues with rewrites() and undertow-handlers.conf.
+ * <p>
+ * Because the rewrite happens after the initial dispatch this handler detects if
+ * the path has been rewritten and updates the servlet target.
+ *
+ * This is a bit of a hack, it needs a lot more thinking about a clean way to handle
+ * this
+ */
+public class RewriteCorrectingHandlerWrappers {
+
+    private static final AttachmentKey<String> OLD_RELATIVE_PATH = AttachmentKey.create(String.class);
+
+    static class PreWrapper implements HandlerWrapper {
+
+        @Override
+        public HttpHandler wrap(final HttpHandler handler) {
+            return new HttpHandler() {
+                @Override
+                public void handleRequest(HttpServerExchange exchange) throws Exception {
+                    exchange.putAttachment(OLD_RELATIVE_PATH, exchange.getRelativePath());
+                    handler.handleRequest(exchange);
+                }
+            };
+        }
+    }
+
+    static class PostWrapper implements HandlerWrapper {
+        @Override
+        public HttpHandler wrap(final HttpHandler handler) {
+            return new HttpHandler() {
+                @Override
+                public void handleRequest(HttpServerExchange exchange) throws Exception {
+                    String old = exchange.getAttachment(OLD_RELATIVE_PATH);
+                    if (!old.equals(exchange.getRelativePath())) {
+                        ServletRequestContext src = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
+                        ServletPathMatch info = src.getDeployment().getServletPaths()
+                                .getServletHandlerByPath(exchange.getRelativePath());
+                        src.setCurrentServlet(info.getServletChain());
+                        src.setServletPathMatch(info);
+                    }
+                    handler.handleRequest(exchange);
+                }
+            };
+        }
+    }
+
+}

--- a/extensions/undertow/runtime/src/main/java/io/quarkus/undertow/runtime/UndertowHandlersConfServletExtension.java
+++ b/extensions/undertow/runtime/src/main/java/io/quarkus/undertow/runtime/UndertowHandlersConfServletExtension.java
@@ -1,0 +1,51 @@
+package io.quarkus.undertow.runtime;
+
+import java.io.InputStream;
+import java.util.List;
+
+import javax.servlet.ServletContext;
+
+import io.undertow.Handlers;
+import io.undertow.server.HandlerWrapper;
+import io.undertow.server.HttpHandler;
+import io.undertow.server.handlers.builder.PredicatedHandler;
+import io.undertow.server.handlers.builder.PredicatedHandlersParser;
+import io.undertow.servlet.ServletExtension;
+import io.undertow.servlet.api.DeploymentInfo;
+
+/**
+ * Registers the Undertow handlers configured in the app's META-INF/undertow-handlers.conf
+ */
+public class UndertowHandlersConfServletExtension implements ServletExtension {
+
+    private static final String META_INF_UNDERTOW_HANDLERS_CONF = "META-INF/undertow-handlers.conf";
+
+    public static boolean existsConfFile() {
+        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+        return classLoader.getResource(META_INF_UNDERTOW_HANDLERS_CONF) != null;
+    }
+
+    @Override
+    public void handleDeployment(DeploymentInfo deploymentInfo, ServletContext servletContext) {
+        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+        InputStream handlers = classLoader.getResourceAsStream(META_INF_UNDERTOW_HANDLERS_CONF);
+        if (handlers != null) {
+            // From Stuart Douglas: Ideally these would be parsed at deployment time and passed into a template,
+            // however they are likely not bytecode serialisable. Even though this approach
+            // does not 100% align with the Quarkus ethos I think it is ok in this case as
+            // the gains would be marginal compared to the cost of attempting to make
+            // every predicate bytecode serialisable.
+            List<PredicatedHandler> handlerList = PredicatedHandlersParser.parse(handlers, classLoader);
+            if (!handlerList.isEmpty()) {
+                deploymentInfo.addOuterHandlerChainWrapper(new RewriteCorrectingHandlerWrappers.PostWrapper());
+                deploymentInfo.addOuterHandlerChainWrapper(new HandlerWrapper() {
+                    @Override
+                    public HttpHandler wrap(HttpHandler handler) {
+                        return Handlers.predicates(handlerList, handler);
+                    }
+                });
+                deploymentInfo.addOuterHandlerChainWrapper(new RewriteCorrectingHandlerWrappers.PreWrapper());
+            }
+        }
+    }
+}

--- a/extensions/undertow/runtime/src/main/java/io/quarkus/undertow/runtime/UndertowHandlersConfServletExtension.java
+++ b/extensions/undertow/runtime/src/main/java/io/quarkus/undertow/runtime/UndertowHandlersConfServletExtension.java
@@ -18,12 +18,7 @@ import io.undertow.servlet.api.DeploymentInfo;
  */
 public class UndertowHandlersConfServletExtension implements ServletExtension {
 
-    private static final String META_INF_UNDERTOW_HANDLERS_CONF = "META-INF/undertow-handlers.conf";
-
-    public static boolean existsConfFile() {
-        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
-        return classLoader.getResource(META_INF_UNDERTOW_HANDLERS_CONF) != null;
-    }
+    public static final String META_INF_UNDERTOW_HANDLERS_CONF = "META-INF/undertow-handlers.conf";
 
     @Override
     public void handleDeployment(DeploymentInfo deploymentInfo, ServletContext servletContext) {

